datafusion/physical-expr/src/aggregate/utils.rs

```diff
@@ -37,45 +37,107 @@ pub fn get_accum_scalar_values_as_arrays(
         .collect::<Vec<_>>())
 }
 
-pub fn calculate_result_decimal_for_avg(
```

Comment:
This function is called once per output row in the current code. I want to be able to perform the setup calculations once and then call the minimal code per row in a loop. Thus this refactor

Reply:
Is it called per output row? Isn't it called in `evaluate`? I think `evaluate` is called only to get final aggregate value?

Reply:
>  Isn't it called in evaluate? 

Yes

> I think evaluate is called only to get final aggregate value?

Correct -- it is called *for each group* to get the final aggregate value.  The ASCII art on this comment https://github.com/apache/arrow-datafusion/issues/4973#issuecomment-1608068287 might help to understand why